### PR TITLE
Before hook for onRun of Catalog entities

### LINKED ISSUE #3916: Hook for onRun of Catalog Entities
In order to allow Extensions to hook into the "Run"-Event (e.g. Connect to Cluster) we need a hook in the extension API that is triggered if someone "connects" / "runs" a catalog api entity.

### PR DESCRIPTION
**Pull request by chenhunghan**
- Add extension api `Renderer.Catalog.catalogEntities.addOnRunHook`

Can be used like
```tsx
Renderer.Catalog.catalogEntities.addOnRunHook(catalogEntityUid, (entity) => {
   // return true => invoke onRun, false => don't invoke onRun;
   return true;
});
```

- Add `prettier` to be able to use `.toMatchInlineSnapshot()`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -357,6 +357,7 @@
     "postcss": "^8.3.6",
     "postcss-loader": "4.3.0",
     "postinstall-postinstall": "^2.1.0",
+    "prettier": "^2.4.1",
```
Comment (Nokel81) [inline package.json:360]: Why is this added?

File: src/common/catalog/catalog-entity.ts [hunk 2]
```diff
@@ -230,7 +231,21 @@ export abstract class CatalogEntity<
     return this.metadata.name;
   }
 
+  /**
+	 * Trigger when user click on the icon in details panel
```
Comment (chenhunghan) [inline src/common/catalog/catalog-entity.ts]: not sure why github renders like this, this looks normal (without extra indents) in my VSCode.

File: src/renderer/components/+catalog/catalog-entity.store.tsx [hunk 3]
```diff
@@ -19,103 +19,13 @@
...
 import { autoBind } from "../../../common/utils";
-import { Badge } from "../badge";
-import { navigation } from "../../navigation";
-import { searchUrlParam } from "../input";
-import { makeCss } from "../../../common/utils/makeCss";
-import { KubeObject } from "../../../common/k8s-api/kube-object";
-
-const css = makeCss(styles);
-
-export class CatalogEntityItem<T extends CatalogEntity> implements ItemObject {
```
Comment (chenhunghan) [inline src/renderer/components/+catalog/catalog-entity.store.tsx:39]: Move `CatalogEntityItem` to a separate file (easier to mock in tests)

File: src/common/catalog/catalog-entity.ts [hunk 4]
```diff
@@ -230,7 +231,21 @@ export abstract class CatalogEntity<
...
   public abstract onRun?(context: CatalogEntityActionContext): void | Promise<void>;
   public abstract onContextMenuOpen(context: CatalogEntityContextMenuContext): void | Promise<void>;
+
+	/**
+	 * Trigger when user click on the icon in details panel
+	 * 
+	 * @remarks
+	 * Priority > this.onRun, if presents, onRun won't be called.
+	 */
+	public abstract onClickDetailIcon?(context: CatalogEntityActionContext): void;
```
Comment (Nokel81) [inline src/common/catalog/catalog-entity.ts]: I disagree that this is the best API for this, especially since https://github.com/lensapp/lens/issues/3909 has now been asked and because I assume you want to do the work before any `onRun` call.

Furthermore, "detail icon" is too much of an implementation detail and locks us in one design.

File: src/main/catalog-sources/weblinks.ts [hunk 5]
```diff
@@ -75,7 +75,9 @@ export function syncWeblinks() {
             uid: weblinkData.id,
             name: weblinkData.name,
             source: "local",
-            labels: {}
+            labels: {
+              triggerOnRunAfterOnClickDetailIcon: "true",
+            }
```
Comment (Nokel81) [inline src/main/catalog-sources/weblinks.ts]: I highly disagree that we should use labels as a means for doing this sort of logic.

File: src/common/catalog-entities/web-link.ts [hunk 6]
```diff
@@ -46,6 +46,12 @@ export class WebLink extends CatalogEntity<CatalogEntityMetadata, WebLinkStatus,
     return;
   }
 
+  onClickDetailIcon(context: CatalogEntityActionContext) {
+    catalogCategoryRegistry
+      .getCategoryForEntity<WebLinkCategory>(this)
+      ?.emit("onClickDetailIcon", this, context);
+  }
```
Comment (Nokel81) [inline src/common/catalog-entities/web-link.ts]: From talking to @msa0311 I was under the impression that we were wanting to block the "onRun" if any of these pre-hooks calls the equivalent of `preventDefault` which this form cannot support because the handlers are run concurrently.

**Review by Nokel81 — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (chenhunghan) [reply package.json:360]: To be able to use `matchInlineSnapshot()`

File: src/main/catalog-sources/weblinks.ts (hunk 5 above)
Comment (chenhunghan) [reply src/main/catalog-sources/weblinks.ts]: Yeah, just noticed labels will show up in the UI, so I guess we need to use other

File: src/common/catalog/catalog-entity.ts (hunk 2 above)
Comment (Nokel81) [reply src/common/catalog/catalog-entity.ts]: Are you using tab characters?

File: src/main/catalog-sources/weblinks.ts (hunk 5 above)
Comment (chenhunghan) [reply src/main/catalog-sources/weblinks.ts]: Move under `.spec`

File: src/renderer/components/+catalog/catalog-entity-details.tsx [hunk 7]
```diff
@@ -75,7 +75,7 @@ export class CatalogEntityDetails<T extends CatalogEntity> extends Component<Pro
                     item.onRun(catalogEntityRunContext);
                   }
 
-                  if (typeof item.entity.onClickDetailIcon === "function" && Boolean(item.entity.metadata?.labels?.triggerOnRunAfterOnClickDetailIcon)) {
+                  if (typeof item.entity.onClickDetailIcon === "function" && item.entity.spec?.triggerOnRunAfterOnClickDetailIcon) {
                     item.onRun(catalogEntityRunContext);
```
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-details.tsx]: 1. This should probably be in the above `if` so as to not check twice.
2. I don't think this does what you want it to do because `onClickDetailIcon` is very much an asyncronous function which will return immediately. I think it would be better to have a "beforeOnRun" hook that gets passed a type that can be `preventDefault`-ed.

Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: 1. It's a bit hard to explain why order like this, maybe it's easier to look at the test case "prioritize onClickDetailIcon over onRun, but always trigger onRun if triggerOnRunAfterOnClickDetailIcon='true'" and "prioritize onClickDetailIcon over onRun"

"prioritize onClickDetailIcon over onRun" is a test case for
```tsx
if (typeof item.entity.onClickDetailIcon === "function") {
    item.onClickDetailIcon(catalogEntityRunContext);
  }	else {
    item.onRun(catalogEntityRunContext);
  }
```
and
"prioritize onClickDetailIcon over onRun, but always trigger onRun if triggerOnRunAfterOnClickDetailIcon='true'" is for this part
```tsx
if (typeof item.entity.onClickDetailIcon === "function" && item.entity.spec?.triggerOnRunAfterOnClickDetailIcon) {
    item.onRun(catalogEntityRunContext);
  }
```

2. `onClickDetailIcon` is a sync function (see typings) that doesn't return. What you describe: "beforeOnRun" + preventDefaulted equals to having `onClickDetailIcon` and set `triggerOnRunAfterOnClickDetailIcon=false`

Comment (Nokel81) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: It is a sync function yes, but you are doing an "emit" which, while sync. Returns immediately and not after all the listeners return.

Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: Not sure I understand, can you elaborate, like which functions return?

**Comment by jim-docker** [comment]
I don't understand the need for this but instead of adding `onClickDetailIcon` can't you add something to the run context that lets `onRun` determine if the icon was clicked? Breaking change?

File: package.json (hunk 1 above)
Comment (Nokel81) [reply package.json:360]: It is a peer dependency then?

File: src/renderer/components/+catalog/catalog-entity-details.tsx (hunk 7 above)
Comment (Nokel81) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: Take https://github.com/lensapp/lens/pull/3911/files/aa70ca4154f945c0ee369167e4826248108856b7#diff-5099912721d6845c3d29e2acaa3fd4a50285eec049fd7aa76e8f025e363b4b2bR49-R53 for instance, any method that has been registered with the Weblinks category will be added to the JS event loop and then the `emit` call will return immediately but because of how the event loop works the return will happen and then continue executing before any of those listeners (or handlers) are executed.

**Comment by Nokel81** [comment]
> I don't understand the need for this but instead of adding `onClickDetailIcon` can't you add something to the run context that lets `onRun` determine if the icon was clicked? Breaking change?

That would also be a possible solution. However, according to @msa0311 this is desirable to happen on attempts for `onRun`. So from that POV the name of even this proposed function is wrong.

File: package.json (hunk 1 above)
Comment (chenhunghan) [reply package.json:360]: We need prettier in dependencies because we are using jest v26. v27 doesn't need that. See https://github.com/facebook/jest/pull/11566
And https://github.com/facebook/jest/issues/8467

File: src/renderer/components/+catalog/catalog-entity-details.tsx (hunk 7 above)
Comment (panuhorsmalahti) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: That's not true, at least for Node.js events and the "events" package for npm. (Not sure which eventemitter we're using here).

```
var EventEmitter = require('events')
 
var ee = new EventEmitter()
ee.on('message', function () {
  console.log("2");
})
console.log("1");
ee.emit('message', 'hello world')
console.log("3");
```

This prints 1, 2 and then 3.

Node.js docs: "When the EventEmitter object emits an event, all of the functions attached to that specific event are called synchronously. Any values returned by the called listeners are ignored and discarded."

File: src/common/catalog/catalog-entity.ts (hunk 2 above)
Comment (chenhunghan) [reply src/common/catalog/catalog-entity.ts]: Thanks, fixed.

**Comment by chenhunghan** [comment]
> I don't understand the need for this but instead of adding `onClickDetailIcon` can't you add something to the run context that lets `onRun` determine if the icon was clicked? Breaking change?

Because we need to add the hooks to onRun via extension, from extension there is no way of changing `onRun`...

File: src/renderer/components/+catalog/catalog-entity-details.tsx (hunk 7 above)
Comment (Nokel81) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: okay fair enough. However async function return immediately anyway so if any they aren't awaited.

File: src/common/catalog-entities/general.ts [hunk 8]
```diff
@@ -39,6 +39,10 @@ export class GeneralEntity extends CatalogEntity<CatalogEntityMetadata, CatalogE
     navigate(this.spec.path);
   }
 
+  onClickDetailIcon() {
```
Comment (Nokel81) [inline src/common/catalog-entities/general.ts]: Still don't like this name.

File: src/common/catalog-entities/kubernetes-cluster.ts [hunk 9]
```diff
@@ -52,6 +52,7 @@ export interface KubernetesClusterSpec extends CatalogEntitySpec {
     background?: string;
   };
   accessibleNamespaces?: string[];
+	triggerOnRunAfterOnClickDetailIcon?: boolean
```
Comment (Nokel81) [inline src/common/catalog-entities/kubernetes-cluster.ts]: I think this should just happen

File: src/common/catalog-entities/general.ts (hunk 8 above)
Comment (chenhunghan) [reply src/common/catalog-entities/general.ts]: Just rebased to master, no new commits yet :)

File: src/renderer/components/+catalog/catalog-entity-details.tsx (hunk 7 above)
Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: Not sure about that too, but I guess it's out of scope of this PR..

File: src/renderer/components/+catalog/catalog-entity.store.tsx (hunk 3 above)
Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity.store.tsx:39]: Do it in https://github.com/lensapp/lens/pull/3927

File: src/common/catalog-entities/kubernetes-cluster.ts (hunk 9 above)
Comment (chenhunghan) [reply src/common/catalog-entities/kubernetes-cluster.ts]: Removed

File: src/extensions/renderer-api/catalog.ts [hunk 10]
```diff
@@ -48,6 +48,33 @@ export class CatalogEntityRegistry {
...
+
+  /**
+   * Returns all catalog entities' onRun hooks
+   */
+  getAllOnRunHooks(): Array<{
+    catalogEntityUid: CatalogEntity["metadata"]["uid"];
+    onRunHook: CatelogEntityOnRunHook;
+  }> {
+    return registry.getAllOnRunHooks();
+  }
```
Comment (Nokel81) [inline src/extensions/renderer-api/catalog.ts]: I don't think we need this API.

File: src/extensions/renderer-api/catalog.ts [hunk 11]
```diff
@@ -48,6 +48,33 @@ export class CatalogEntityRegistry {
...
+    return registry.addOnRunHook(catalogEntityUid, onRunHook);
+  }
+
+  /**
+   * Returns one catalog entity onRun hook by catalog entity uid
+   */
+  getOnRunHook(catalogEntityUid: CatalogEntity["metadata"]["uid"]): CatelogEntityOnRunHook | undefined {
+    return registry.getOnRunHook(catalogEntityUid);
+  }
+
```
Comment (Nokel81) [inline src/extensions/renderer-api/catalog.ts]: We should support several onRunHooks. You sort of do already but not very well since you only ever return the first hook.

File: src/renderer/api/catalog-entity-registry.ts [hunk 12]
```diff
@@ -27,15 +27,25 @@ import type { Cluster } from "../../main/cluster";
 import { ClusterStore } from "../../common/cluster-store";
 import { Disposer, iter } from "../utils";
 import { once } from "lodash";
+import type { CatalogEntityItem } from "../../renderer/components/+catalog/catalog-entity-item";
 
 export type EntityFilter = (entity: CatalogEntity) => any;
+export type CatelogEntityOnRunHook = (entity: CatalogEntity | CatalogEntityItem<CatalogEntity>) => boolean | Promise<boolean>;
```
Comment (Nokel81) [inline src/renderer/api/catalog-entity-registry.ts]: spelling mistake: should be catalog...

Comment (Nokel81) [reply src/renderer/api/catalog-entity-registry.ts]: Also I think that leaking `CatalogEnityItem` to the extension API is a very bad idea.

File: src/renderer/api/catalog-entity-registry.ts [hunk 13]
```diff
@@ -27,15 +27,25 @@ import type { Cluster } from "../../main/cluster";
...
+
+type CatalogEntityUid = CatalogEntity["metadata"]["uid"];
```
Comment (Nokel81) [inline src/renderer/api/catalog-entity-registry.ts:36]: I would just use `string` directly.

File: src/renderer/api/catalog-entity-registry.ts [hunk 14]
```diff
@@ -27,15 +27,25 @@ import type { Cluster } from "../../main/cluster";
...
   protected _entities = observable.map<string, CatalogEntity>([], { deep: true });
   protected filters = observable.set<EntityFilter>([], {
     deep: false,
   });
+  protected entityOnRunHooks = observable.set<{
+    catalogEntityUid: CatalogEntityUid;
+    onRunHook: CatelogEntityOnRunHook
+  }>([], {
+    deep: false,
+  });
```
Comment (Nokel81) [inline src/renderer/api/catalog-entity-registry.ts]: This is a very strange type, why is it not `observable.map<string, CatalogEnitytOnRunHook[]>([], { deep: false });`?

File: src/renderer/components/+catalog/catalog-entity-details.tsx [hunk 15]
```diff
@@ -68,8 +68,12 @@ export class CatalogEntityDetails<T extends CatalogEntity> extends Component<Pro
                 material={item.entity.spec.icon?.material}
                 background={item.entity.spec.icon?.background}
                 disabled={!item?.enabled}
-                onClick={() => item.onRun(catalogEntityRunContext)}
-                size={128} />
+                onClick={() => {
+                  this.props.onClickDetailPanelIcon(item);
```
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-details.tsx]: I don't think we should pass this in as props, we should have an unexported function that takes an `CatalogEntity` and does the work once (DRY as it were).

File: src/renderer/components/+catalog/catalog-entity-item.tsx [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    if (!onRunHook) {
+      this.entity.onRun(ctx);
+
+      return;
+    }
+
+    if (typeof onRunHook === "function") {
+      // if onRunHook() returns a Promise, we wait for it to resolve
+      // if not, just take whatever it returns
+      Promise.resolve(onRunHook(toJS(this.entity))).then((shouldRun) => {
```
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-item.tsx]: This can lead to `UnhandledPromiseRejectionError` which is bad design.

File: src/renderer/components/+catalog/catalog-entity-item.tsx [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  onRun(onRunHook: CatelogEntityOnRunHook | undefined, ctx: CatalogEntityActionContext) {
+    if (!onRunHook) {
+      this.entity.onRun(ctx);
+
+      return;
+    }
+
+    if (typeof onRunHook === "function") {
```
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog-entity-item.tsx]: These two checks can be combined:

```js
if (typeof onRunHook !== "function") {
  ...
  return;
}

...
```

File: src/extensions/renderer-api/catalog.ts [hunk 18]
```diff
@@ -48,6 +48,33 @@ export class CatalogEntityRegistry {
...
     return registry.getItemsForCategory(category);
   }
+
+  /**
+   * Add a onRun hook to a catalog entity.
+   * @param catalogEntityUid The uid of the catalog entity
+   * @param onRunHook The function that should return a boolean if the onRun of catalog entity should be triggered.
+   * @returns A function to remove that hook
+   */
+  addOnRunHook(catalogEntityUid: CatalogEntity["metadata"]["uid"], onRunHook: CatelogEntityOnRunHook) {
```
Comment (Nokel81) [inline src/extensions/renderer-api/catalog.ts]: I think this should be instead called `onBeforeRun` so as to be more descriptive of when it will be run.

File: src/extensions/renderer-api/catalog.ts (hunk 10 above)
Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: Can you explain why? Intention: allow extension developer to check if hooks are already added.

File: src/renderer/components/+catalog/catalog-entity-item.tsx (hunk 17 above)
Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity-item.tsx]: Maybe, but with less readability IMHO.

File: src/extensions/renderer-api/catalog.ts (hunk 10 above)
Comment (Nokel81) [reply src/extensions/renderer-api/catalog.ts]: Why would it be the extension developer care? We should just use a `Set` then.

Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: At least in space extension we need this to check when the clusters are synced from the cloud.

Comment (Nokel81) [reply src/extensions/renderer-api/catalog.ts]: They should just add the handler again (using the built-in set type logic) or you should add a method to do that check. I don't see the point of this method in particular.

Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: realised we had `getOnRunHook` already, we can remove this.

File: src/extensions/renderer-api/catalog.ts (hunk 11 above)
Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: I think this should be another PR,

File: src/renderer/api/catalog-entity-registry.ts (hunk 12 above)
Comment (chenhunghan) [reply src/renderer/api/catalog-entity-registry.ts]: Thanks, fix the typo. About `CatalogEnityItem`, it more about reflecting the true types... but I guess we can remove it.

File: src/renderer/api/catalog-entity-registry.ts (hunk 13 above)
Comment (chenhunghan) [reply src/renderer/api/catalog-entity-registry.ts:36]: IMHO this is more descriptive we are using uid from `catalog.metadata.uid`

File: src/extensions/renderer-api/catalog.ts (hunk 18 above)
Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: If we want to support multiple hooks in the future, `onBeforeRun` is less descriptive IMO.

File: src/renderer/api/catalog-entity-registry.ts (hunk 14 above)
Comment (chenhunghan) [reply src/renderer/api/catalog-entity-registry.ts]: We need `catalogEntityUid` because there are `getOnRunHook `

File: src/renderer/components/+catalog/catalog-entity-item.tsx (hunk 16 above)
Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity-item.tsx]: `UnhandledPromiseRejectionError` is only in node.js not in React, right? Good point about the error handling though, added some more test cases https://github.com/lensapp/lens/pull/3911/commits/24f5f9950afb432c98c815387c9f41786f1ea667

File: src/renderer/components/+catalog/catalog-entity-details.tsx (hunk 15 above)
Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity-details.tsx]: Can the refactoring be another PR?

File: src/renderer/components/icon/icon.tsx [hunk 19]
```diff
@@ -114,14 +114,14 @@ export class Icon extends React.PureComponent<IconProps> {
     };
 
     // render as inline svg-icon
-    if (svg) {
+    if (typeof svg === "string") {
```
Comment (chenhunghan) [inline src/renderer/components/icon/icon.tsx:117]: This is here just for easier mocking in tests (less crashes in components)

File: src/extensions/lens-renderer-extension.ts [hunk 20]
```diff
@@ -77,7 +77,7 @@ export class LensRendererExtension extends LensExtension {
   }
 
   /**
-   * Add a filtering function for the catalog catogries. This will be removed if the extension is disabled.
+   * Add a filtering function for the catalog categories. This will be removed if the extension is disabled.
```
Comment (chenhunghan) [inline src/extensions/lens-renderer-extension.ts:80]: Was a typo.

File: src/extensions/renderer-api/catalog.ts (hunk 11 above)
Comment (Nokel81) [reply src/extensions/renderer-api/catalog.ts]: No because this will have to be supported until the next major

File: src/renderer/api/catalog-entity-registry.ts (hunk 13 above)
Comment (Nokel81) [reply src/renderer/api/catalog-entity-registry.ts:36]: Ok

File: src/renderer/api/catalog-entity-registry.ts (hunk 14 above)
Comment (Nokel81) [reply src/renderer/api/catalog-entity-registry.ts]: The top level map's key would be `CatalogEntityUid`

File: src/renderer/components/+catalog/catalog-entity-item.tsx (hunk 16 above)
Comment (Nokel81) [reply src/renderer/components/+catalog/catalog-entity-item.tsx]: no that is part of the JS spec. If you open up devtools and type `Promise.reject(4)` into the console you should see something like this. Sure it isn't the same name but it is still bad design
Screen Shot 2021-10-05 at 8 42 11 AM

File: src/extensions/renderer-api/catalog.ts (hunk 18 above)
Comment (Nokel81) [reply src/extensions/renderer-api/catalog.ts]: How is it less descriptive? It better shows when we plan on running them

Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: Well, it's just my opinion, if you think "onBeforeRun" is better, I would just change it

File: src/renderer/components/+catalog/catalog-entity-item.tsx (hunk 16 above)
Comment (chenhunghan) [reply src/renderer/components/+catalog/catalog-entity-item.tsx]: It has been fixed.

File: src/extensions/renderer-api/catalog.ts (hunk 18 above)
Comment (Nokel81) [reply src/extensions/renderer-api/catalog.ts]: I meant the argument would be `onBeforeRun` and this method would be `addOnBeforeRun`

File: src/renderer/api/catalog-entity-registry.ts (hunk 14 above)
Comment (chenhunghan) [reply src/renderer/api/catalog-entity-registry.ts]: `string` is not as obvious as `catalogEntityUid`, but again, just my opinion, maybe you know more about the convention in Lens.

File: src/extensions/renderer-api/catalog.ts (hunk 18 above)
Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: OK, `addOnBeforeRun` makes sense.

File: src/renderer/api/catalog-entity-registry.ts (hunk 14 above)
Comment (Nokel81) [reply src/renderer/api/catalog-entity-registry.ts]: We don't really have a convention... (namely we use both styles). What I meant was I should have suggested the type `observable.map<CatalogEntityUid, IObservableSet<CatelogEntityOnRunHook>>`

File: src/extensions/renderer-api/catalog.ts (hunk 18 above)
Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: Edited https://github.com/lensapp/lens/pull/3911/commits/74eb89c361ceff7f396f405ddbf7eb08e4da3b99

File: src/extensions/renderer-api/catalog.ts (hunk 11 above)
Comment (chenhunghan) [reply src/extensions/renderer-api/catalog.ts]: At least in current timeframe I don't think I can do this before 5.3. @Nokel81 would you like to to take over on the multiple hooks part?

File: src/renderer/api/catalog-entity-registry.ts (hunk 14 above)
Comment (chenhunghan) [reply src/renderer/api/catalog-entity-registry.ts]: OK, thanks for clarify.

Comment (chenhunghan) [reply src/renderer/api/catalog-entity-registry.ts]: Changed https://github.com/lensapp/lens/pull/3911/commits/c8631346ca0d734c233a97d3134f1a9362b6900f

**Review by jim-docker — COMMENTED** [review]
reviewed @Nokel81 's commits, LGTM but my head is spinning. @Nokel81 should give the final approval...

File: src/extensions/renderer-api/catalog.ts [hunk 21]
```diff
@@ -50,21 +51,14 @@ export class CatalogEntityRegistry {
   }
 
   /**
-   * Add a onBeforeRun hook to a catalog entity.
+   * Add a onRun hook to a catalog entity. If `onBeforeRun` was previously added then it will not be added again
```
Comment (jim-docker) [inline src/extensions/renderer-api/catalog.ts]: Why change to `onRun`? ("Add a onRun hook to...")

Comment (Nokel81) [reply src/extensions/renderer-api/catalog.ts]: Good catch, I actually copied this over from the non-exposed registry. So I have fixed it there too.

**Review by Nokel81 — APPROVED** [review]
